### PR TITLE
Add option to switch cms account being used

### PR DESCRIPTION
**Pull request by hirthy**
Please take a look @smoll, not sure if this is the right approach. I'd like to clean up switch_accounts_page in the future.

### CODE REVIEW

File: pages/youtube/switch_accounts_page.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+module YouTube
+  class SwitchAccountsPage < BasePage
+
+    def switch_accounts(account)
+      browser.button(:class => 'yt-masthead-user-icon').click
```
Comment (smoll) [inline pages/youtube/switch_accounts_page.rb]: The fundamental structure of this looks okay, but I would probably use page-object accessors for these, like: https://github.com/smoll/erdl/blob/master/pages/youtube/dashboard_page.rb#L3

That lets you DRY up all these button/div definitions in this method body.

File: pages/youtube/routes.rb [hunk 2]
```diff
@@ -2,13 +2,22 @@ module YouTube
...
-        [YouTube::DownloadReportsPage]
-      ]
+      if opts.has_key?('accounts')
+        @default = [
+          [YouTube::LoginPage, :login],
+          [YouTube::SwitchAccountsPage, :switch_accounts, opts[:account]],
+          [YouTube::DashboardPage, :reports],
+          [YouTube::DownloadReportsPage]
+        ]
+      else
```
Comment (smoll) [inline pages/youtube/routes.rb]: Is it better to subclass this and override @default in the constructor:

```ruby
module YouTube
  class SwitchAccounts < Routes
    @default = # your new route
  end
end
```

Then in url_fetcher.rb:

```ruby
# ...
if opts.has_key?('accounts')
  route = class_from("#{opts[:site]}::SwitchAccounts")
else
  route = class_from("#{opts[:site]}::Routes").new(@browser,opts)
end
```

Not sure, not sure which of these are more easily unit-tested. From my first pass at it, I've already noticed PageObject is quite difficult to stub: https://github.com/smoll/erdl/pull/3/files#r26880197

Comment (smoll) [inline pages/youtube/routes.rb]: On second thought, no, your solution seems simpler. We could even initialize `@default` using instance methods, like:

```ruby
# routes.rb

    def initialize(browser, opts = {})
      @browser = browser # needed by PageObject::PageFactory
      @default = opts.has_key?('accounts') ? switch_account_route : default_route
      @first = @default.first[0] # 0th param is page class, 1st param is method
      @last = @default.last[0]
      set_default
    end

    def default_route
      [
          [YouTube::LoginPage, :login],
          [YouTube::DashboardPage, :reports],
          [YouTube::DownloadReportsPage]
      ]
    end

    def switch_account_route
      [
          [YouTube::LoginPage, :login],
          [YouTube::SwitchAccountsPage, :switch_accounts, opts[:account]],
          [YouTube::DashboardPage, :reports],
          [YouTube::DownloadReportsPage]
      ]
    end
```

Comment (smoll) [inline pages/youtube/routes.rb]: I've added a bunch of unit tests and standardized all the constructors in a pretty uniform way I think -- please rebase and ensure unit tests pass and we can get this merged :)

File: pages/youtube/switch_accounts_page.rb [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+      browser.div(:id => 'yt-delegate-accounts').wait_until_present
+      if browser.div(:class => 'yt-masthead-picker-name').text == account
+        puts 'already on desired account'
+        return
+      else
+        puts 'finding desired account...'
+        browser.div(:id => 'yt-delegate-accounts').links(:class => 'yt-masthead-account-picker-option').each do |link|
```
Comment (gian-orchard) [inline pages/youtube/switch_accounts_page.rb]: Once you have a PO element defined for the links, you should be able to do a find on its child element and click it directly:

`name_of_the_collection_of_links_elements.find { |link| link.div_element(class: 'yt-masthead-picker-name').text == account }.click`

**Comment by hirthy** [comment]
I've refactored the code with both of your suggestions and the unit tests pass. However, the tests don't cover the switch_accounts method and I've had some trouble getting those lines to be covered.

**Comment by smoll** [comment]
I'll add a separate PR to ignore page classes in coveralls' coverage reports -- there's almost no probative value in unit testing them (because if the site, e.g. YouTube, changes its frontend our unit tests will continue to pass but this tool will no longer work)...

**Pull request merged**